text/0000-unified_coroutines.md

```diff
@@ -0,0 +1,411 @@
+
+Alternatives:
+- Only current alternative is storing required information inside side channels such as `Rc<RefCell<Args>>` 
+or a thread local storage, which introduces runtime overhead and requires `std`. Proposed design is `no_std` compatible.
+- The proposed syntax could be changed, but from the explored options, we pose that the simplest syntax is best, even though it introduces new semantics.
+- Leave the generator as is, leaving it disconnected from the rest of the language.
+- Remove generators completely 
+
+# Prior art
+[prior-art]: #prior-art
+
+- Pre-RFC published by different author on rust-internals forum [link](https://internals.rust-lang.org/t/pre-rfc-generator-resume-args/10011/5)
+   
+  Explored the design space and proposed a basic design of the modified `Generator trait`. The described approach solved only the 'Generator resume arguments' part of this RFC, and did not attempt to unify generators with closures, which resulted in unnecessarily complex design, which kept the generators as a separate concept, and even introduced another layer of complexity in form of a trait alias for generators with/without arguments. But nonetheless, reading this discussion was an invaluable source of ideas in this design space.
+  
+- Work on implementing futures which wouldn't require TLS: [link](https://github.com/rust-lang/rust/issues/62918)
+
+- Python & Lua coroutines - They can be resumed with arguments, with yield expression returning these values [usage](https://www.tutorialspoint.com/lua/lua_coroutines.htm). 
+  
+  These are interesting, since they both adopt a syntax, in which the yield expression returns values passed to resume. We think that this approach is right one for dynamic languages like Python or lua but the wrong one for Rust. The reason is, these languages are dynamically typed, and allow passing of multiple values into the coroutine. The design proposed here is static, and allows passing only a single argument into the coroutine, a tuple. The argument tuple is treated the same way as in the `Fn*` family of traits. 
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Proposed syntax: Do we somehow require assignemnt from yield expression(As outlined by [different pre-rfc](https://internals.rust-lang.org/t/pre-rfc-generator-resume-args/10011)), or we do we specify arguments only at the start of the coroutine, and require 
+explanation of the different behavior in combination with the `yield` keyword explanation ?
+
+- Do we unpack the coroutine arguments, unifying the behavior with closures, or do we force only a single argument and encourage the use of tuples ?
+
+- Do we allow non `'static` coroutine arguments ? How would they interact with the lifetime of the generator, if the generator moved the values passed into `resume` into its local state ?
```

Comment:
If we do allow non-`'static` coroutine arguments, some use cases will want arguments that live only for a single call to `resume`. In this case, the syntactic approach proposed here doesn't work very well- the arguments would *change lifetime* (and thus type!) throughout the execution of the generator.

If instead the arguments are provided as the value of a `yield` expression, each one could have a separate set of lifetimes limited by the actual live range of the expression's result. This fits much more naturally with how lifetimes work in non-generator code, especially with NLL.

Reply:
Yes, this is the primary issue with my approach. 

The generators in this case assume lifetime issues are not resolved at the source code / AST level, but rather at the dataflow / MIR level.

This means that in this case:

```rust
1. |a| {
2.    loop {
3.       println!("a : {:?}", a);
4.       let (a,) = yield 1;
5.       println!("b : {:?}", a);
6.       let (a,) = yield 2;
7.    }        
8. }
```
The lifetime of `a`  both starts at the line 6 and ends at the line 4, and starts at line 1, and ends at line 4 depending on the entry point. But, these are 2 different values of `a`, the behavior should be similar to the behavior which would be observed if we have written the generator by hand as a match statment, just like in the RFC.

I was under the assumption that the borrowck is MIR based, not AST based, and therefore this approach would be usable.

Reply:
This is a requirement for one of the primary usecases: futures. They will need to take `Args = (&mut core::task::Context<'_>,)` which contains two lifetimes that are only valid during the current `resume` call.